File: backend/app/policies/policy_engine.py
from __future__ import annotations

from fastapi import HTTPException, status

from app.core.config import get_settings
from app.policies.permissions import resolve_permission_with_fallback
from app.policies.schemas import AccessContext, PolicyDecision, PolicyInput


def allow_decision(
    *,
    required_permissions: list[str] | None = None,
    checked_permissions: list[str] | None = None,
    warnings: list[str] | None = None,
    scope: dict[str, object] | None = None,
) -> PolicyDecision:
    return PolicyDecision(
        allowed=True,
        code="ALLOWED",
        message="Islem baslatilabilir.",
        warnings=warnings or [],
        required_permissions=required_permissions or [],
        checked_permissions=checked_permissions or [],
        scope=scope or {},
    )


def deny_decision(
    code: str,
    message: str,
    reasons: list[str] | None = None,
    *,
    required_permissions: list[str] | None = None,
    checked_permissions: list[str] | None = None,
    warnings: list[str] | None = None,
    scope: dict[str, object] | None = None,
) -> PolicyDecision:
    return PolicyDecision(
        allowed=False,
        code=code,
        message=message,
        reasons=reasons or [message],
        warnings=warnings or [],
        required_permissions=required_permissions or [],
        checked_permissions=checked_permissions or [],
        scope=scope or {},
    )


def _has_permission(context: AccessContext, required_permission: str) -> tuple[bool, list[str]]:
    accepted = resolve_permission_with_fallback(required_permission)
    permissions = set(context.permissions)
    if "*" in permissions or "system.admin" in permissions:
        return True, accepted
    return bool(permissions.intersection(accepted)), accepted


def _scope_denied(context: AccessContext, resource: dict[str, object] | None) -> str | None:
    if not resource:
        return None
    company_id = resource.get("company_id")
    if company_id and not context.company_scope:
        settings = get_settings()
        if settings.effective_auth_required or not settings.is_development:
            return "Bu kayit erisim kapsaminiz disinda."
    if context.company_scope and company_id and str(company_id) not in context.company_scope:
        return "Bu kayit erisim kapsaminiz disinda."
    branch_id = resource.get("branch_id")
    if context.branch_scope and branch_id and str(branch_id) not in context.branch_scope:
        return "Bu kayit erisim kapsaminiz disinda."
    return None


def evaluate_policy(input_data: PolicyInput) -> PolicyDecision:
    context = input_data.context
    if not context.tenant_id:
        return deny_decision(
            "WORKSPACE_CONTEXT_REQUIRED",
            "Calisma alani bilgisi alinamadi.",
        )
    checked_permissions: list[str] = []
    for required in input_data.required_permissions:
        allowed, checked = _has_permission(context, required)
        checked_permissions.extend(checked)
        if not allowed:
            return deny_decision(
                "PERMISSION_DENIED",
                "Bu islemi yapmak icin gerekli yetkiniz bulunmuyor.",
                required_permissions=input_data.required_permissions,
                checked_permissions=checked_permissions,
            )

    scope_reason = _scope_denied(context, input_data.resource)
    if scope_reason:
        return deny_decision(
            "SCOPE_DENIED",
            scope_reason,
            scope={
                "company_scope": context.company_scope or [],
                "branch_scope": context.branch_scope or [],
            },
        )

    record_status = (
        input_data.resource or {}
    ).get("record_status") or context.record_status
    if record_status and input_data.required_record_status:
        if str(record_status) not in input_data.required_record_status:
            return deny_decision(
                "POLICY_DENIED",
                "Bu islem mevcut kayit durumu nedeniyle baslatilamaz.",
                [f"Beklenen kayit durumu: {', '.join(input_data.required_record_status)}"],
            )
    if record_status and str(record_status) in input_data.blocked_record_status:
        return deny_decision(
            "POLICY_DENIED",
            "Bu islem mevcut kayit durumu veya yetkileriniz nedeniyle baslatilamaz.",
        )

    if input_data.resource and bool(input_data.resource.get("is_deleted")):
        return deny_decision("POLICY_DENIED", "Silinmis kayit uzerinde islem yapilamaz.")

    return allow_decision(
        required_permissions=input_data.required_permissions,
        checked_permissions=checked_permissions,
        scope={
            "tenant_id": context.tenant_id,
            "company_id": context.company_id,
            "branch_id": context.branch_id,
        },
    )


def policy_to_http_exception(decision: PolicyDecision) -> HTTPException:
    status_code = status.HTTP_403_FORBIDDEN
    if decision.code == "WORKSPACE_CONTEXT_REQUIRED":
        status_code = status.HTTP_401_UNAUTHORIZED
    if decision.code == "POLICY_DENIED":
        status_code = status.HTTP_409_CONFLICT
    return HTTPException(
        status_code=status_code,
        detail={
            "error": decision.message,
            "code": decision.code,
            "details": decision.model_dump(mode="json"),
            "message": decision.message,
        },
    )


def assert_policy(input_data: PolicyInput) -> PolicyDecision:
    decision = evaluate_policy(input_data)
    if not decision.allowed:
        raise policy_to_http_exception(decision)
    return decision

File: lib/security/policyEngine.ts
// BACKEND_MIGRATION_STATUS: migrate_to_fastapi
// TARGET_BACKEND_MODULE: policy
// TARGET_ENDPOINT: /api/v1/policies/evaluate
// NOTES: Backend authorization source of truth should move to Python Policy Engine.

import 'server-only'

import { NextResponse } from 'next/server'
import { findActionContract } from '@/lib/modules/moduleRegistry'
import { createServiceClient } from '@/lib/supabase/server'
import { AuditLogService } from '@/lib/audit/auditLogService'
import { expandPermissionFallbacks } from './permissionRegistry'
import { hasAnyPermission } from './serverPermissions'
import type { AccessContext } from './accessContext'
import {
  canAccessBranch,
  canAccessCompany,
  canAccessFacility,
  canAccessOrganizationUnit,
  canWriteBranch,
  canWriteCompany,
  canWriteFacility,
  canWriteOrganizationUnit,
  isClosedOrPassive,
  normalizeStatus,
} from './scopePolicy'

export interface PolicyDecision {
  allowed: boolean
  code: string
  message: string
  reasons: string[]
  warnings: string[]
  requiredPermissions?: string[]
  checkedPermissions?: string[]
  scope?: Record<string, any>
}

export interface PolicyInput {
  context: AccessContext
  actionKey: string
  moduleKey?: string
  resourceType?: string
  resourceId?: string
  resource?: Record<string, any>
  requiredPermissions?: string[]
  requiredRecordStatus?: string[]
  blockedRecordStatus?: string[]
  extraRules?: PolicyRule[]
}

export interface PolicyRule {
  key: string
  description?: string
  evaluate(context: AccessContext, input: PolicyInput): PolicyDecision | null | Promise<PolicyDecision | null>
}

export async function evaluatePolicy(input: PolicyInput): Promise<PolicyDecision> {
  const moduleDecision = evaluateModuleAvailability(input)
  if (moduleDecision && !moduleDecision.allowed) return moduleDecision

  const permissionDecision = evaluatePermission(input)
  if (!permissionDecision.allowed) return permissionDecision

  const tenantDecision = evaluateTenantScope(input)
  if (!tenantDecision.allowed) return tenantDecision

  const scopeDecision = await evaluateScope(input)
  if (!scopeDecision.allowed) return scopeDecision

  const statusDecision = evaluateRecordStatus(input)
  if (!statusDecision.allowed) return statusDecision

  for (const rule of input.extraRules || []) {
    const decision = await rule.evaluate(input.context, input)
    if (decision && !decision.allowed) return decision
  }

  return allowDecision({
    warnings: [
      ...(moduleDecision?.warnings || []),
      ...permissionDecision.warnings,
      ...tenantDecision.warnings,
      ...scopeDecision.warnings,
      ...statusDecision.warnings,
      ...(input.extraRules || []).flatMap(() => []),
    ],
    requiredPermissions: permissionDecision.requiredPermissions,
    checkedPermissions: permissionDecision.checkedPermissions,
    scope: scopeDecision.scope,
  })
}

export function allowDecision(overrides: Partial<PolicyDecision> = {}): PolicyDecision {
  return {
    allowed: true,
    code: 'POLICY_ALLOWED',
    message: 'Islem baslatilabilir.',
    reasons: [],
    warnings: [],
    ...overrides,
  }
}

export function denyDecision(
  code: string,
  message: string,
  reasons: string[] = [],
  overrides: Partial<PolicyDecision> = {}
): PolicyDecision {
  return {
    allowed: false,
    code,
    message,
    reasons: reasons.length ? reasons : [message],
    warnings: [],
    ...overrides,
  }
}

export function policyToResponse(decision: PolicyDecision, status = 403) {
  return NextResponse.json({
    error: decision.message || 'Bu islem mevcut kayit durumu veya yetkileriniz nedeniyle baslatilamaz.',
    code: decision.code || 'POLICY_DENIED',
    details: {
      reasons: decision.reasons,
      warnings: decision.warnings,
      required_permissions: decision.requiredPermissions,
      checked_permissions: decision.checkedPermissions,
      scope: decision.scope,
    },
  }, { status })
}

export async function assertPolicy(input: PolicyInput) {
  const decision = await evaluatePolicy(input)
  if (!decision.allowed) {
    await new AuditLogService(createServiceClient()).recordPolicyDenied({
      context: {
        tenantId: input.context.tenantId,
        companyId: input.context.companyId || null,
        branchId: input.context.branchId || null,
        moduleKey: input.moduleKey || input.context.moduleKey || null,
        entityType: input.resourceType || input.context.recordType || null,
        entityId: input.resourceId || input.context.recordId || null,
        userId: input.context.userId || null,
        policyKey: decision.code,
      },
      actionKey: input.actionKey,
      summary: 'Islem is kurali nedeniyle reddedildi.',
      reason: decision.reasons.join(' '),
      metadata: {
        warnings: decision.warnings,
        required_permissions: decision.requiredPermissions,
        checked_permissions: decision.checkedPermissions,
      },
    }).catch(() => null)
  }
  return decision.allowed ? null : policyToResponse(decision, policyStatus(decision))
}

function evaluateModuleAvailability(input: PolicyInput): PolicyDecision | null {
  const actionContract = findActionContract(input.actionKey)
  const moduleKey = input.moduleKey || actionContract?.module.key || input.context.moduleKey
  if (!moduleKey) return null

  const moduleStatus = input.context.moduleStatus?.[moduleKey]
  if (!moduleStatus) {
    return allowDecision({ warnings: [`${moduleKey} modul runtime durumu bulunamadi; mevcut permission davranisi kullanildi.`] })
  }
  if (moduleStatus.status === 'available') {
    return allowDecision({ warnings: moduleStatus.warnings || [] })
  }

  return denyDecision(
    moduleStatus.status === 'disabled'
      ? 'MODULE_DISABLED'
      : moduleStatus.status === 'unlicensed'
        ? 'MODULE_UNLICENSED'
        : moduleStatus.status === 'setup_required'
          ? 'MODULE_SETUP_REQUIRED'
          : 'MODULE_DEPENDENCY_MISSING',
    moduleStatus.blocking_reasons?.[0] || 'Bu modul bu calisma alaninda kullanilabilir degil.',
    moduleStatus.blocking_reasons || [],
    { warnings: moduleStatus.warnings || [] }
  )
}

function evaluatePermission(input: PolicyInput): PolicyDecision {
  const actionContract = findActionContract(input.actionKey)
  const requiredPermissions = input.requiredPermissions
    || [actionContract?.action.permission, actionContract?.action.fallbackPermission].filter(Boolean) as string[]

  if (!requiredPermissions.length) return allowDecision()

  const checkedPermissions = expandPermissionFallbacks(requiredPermissions)
  if (hasAnyPermission(input.context.permissions, requiredPermissions)) {
    return allowDecision({ requiredPermissions, checkedPermissions })
  }

  return denyDecision(
    'PERMISSION_DENIED',
    'Bu islemi yapmak icin gerekli yetkiniz bulunmuyor.',
    ['Yetki kontrolu basarisiz.'],
    { requiredPermissions, checkedPermissions }
  )
}

function evaluateTenantScope(input: PolicyInput): PolicyDecision {
  const resourceTenantId = input.resource?.tenant_id || input.resource?.tenantId
  if (!resourceTenantId || resourceTenantId === input.context.tenantId) return allowDecision()

  return denyDecision(
    'SCOPE_DENIED',
    'Bu kayit erisim kapsaminiz disinda.',
    ['Kayit farkli bir calisma alani kapsaminda.'],
    { scope: { resource_type: input.resourceType, resource_id: input.resourceId || input.resource?.id } }
  )
}

async function evaluateScope(input: PolicyInput): Promise<PolicyDecision> {
  const resourceType = input.resourceType || input.context.recordType
  const resource = input.resource || {}
  const companyId = resource.company_id || resource.companyId || input.context.companyId
  const branchId = resourceType === 'branch' ? input.resourceId || resource.id || input.context.branchId : input.context.branchId
  const organizationUnitId = resourceType === 'organization_unit' ? input.resourceId || resource.id || input.context.organizationUnitId : input.context.organizationUnitId
  const facilityId = resourceType === 'facility' ? input.resourceId || resource.id || input.context.facilityId : input.context.facilityId
  const write = isWritePolicy(input)

  if (branchId) {
    const ok = write ? await canWriteBranch(input.context, branchId) : await canAccessBranch(input.context, branchId)
    if (!ok) return scopeDenied('branch', branchId)
  }
  if (organizationUnitId) {
    const ok = write
      ? await canWriteOrganizationUnit(input.context, organizationUnitId)
      : await canAccessOrganizationUnit(input.context, organizationUnitId)
    if (!ok) return scopeDenied('organization_unit', organizationUnitId)
  }
  if (facilityId) {
    const ok = write ? await canWriteFacility(input.context, facilityId) : await canAccessFacility(input.context, facilityId)
    if (!ok) return scopeDenied('facility', facilityId)
  }
  if (companyId) {
    const ok = write ? await canWriteCompany(input.context, companyId) : await canAccessCompany(input.context, companyId)
    if (!ok) return scopeDenied('company', companyId)
  }

  return allowDecision({
    scope: {
      company_id: companyId || null,
      branch_id: branchId || null,
      organization_unit_id: organizationUnitId || null,
      facility_id: facilityId || null,
    },
  })
}

function evaluateRecordStatus(input: PolicyInput): PolicyDecision {
  const currentStatus = normalizeStatus(input.resource?.record_status || input.resource?.status || input.context.recordStatus)
  const required = (input.requiredRecordStatus || findActionContract(input.actionKey)?.action.requiredRecordStatus || []).map(normalizeStatus)
  const blocked = (input.blockedRecordStatus || findActionContract(input.actionKey)?.action.blockedRecordStatus || []).map(normalizeStatus)

  if (input.resource?.is_deleted === true) {
    return denyDecision('RECORD_DELETED', 'Silinmis kayit uzerinde bu islem baslatilamaz.', ['Kayit silinmis.'])
  }
  if (required.length && !required.includes(currentStatus)) {
    return denyDecision(
      'RECORD_STATUS_BLOCKED',
      'Bu islem mevcut kayit durumunda baslatilamaz.',
      [`Gerekli durum: ${required.join(', ') || '-'}, mevcut durum: ${currentStatus || '-'}.`]
    )
  }
  if (blocked.length && blocked.includes(currentStatus)) {
    return denyDecision(
      'RECORD_STATUS_BLOCKED',
      'Bu islem mevcut kayit durumunda baslatilamaz.',
      [`${currentStatus} durumundaki kayitlar bu isleme kapali.`]
    )
  }
  if (input.resource && isWritePolicy(input) && isClosedOrPassive(input.resource)) {
    return denyDecision(
      'RECORD_STATUS_BLOCKED',
      'Kapali veya pasif kayit uzerinde bu islem baslatilamaz.',
      ['Kayit kapali, pasif veya silinmis gorunuyor.']
    )
  }

  return allowDecision()
}

function scopeDenied(resourceType: string, resourceId?: string | null): PolicyDecision {
  return denyDecision(
    'SCOPE_DENIED',
    'Bu kayit erisim kapsaminiz disinda.',
    ['Kayit tenant, sirket veya alt kapsam yetkinizin disinda.'],
    { scope: { resource_type: resourceType, resource_id: resourceId || null } }
  )
}

function isWritePolicy(input: PolicyInput) {
  const category = findActionContract(input.actionKey)?.action.actionType
  const text = `${input.actionKey} ${category || ''}`.toLocaleLowerCase('tr-TR')
  return ['edit', 'opening', 'closing', 'start', 'update', 'terminate', 'suspend', 'document', 'operation'].some(token => text.includes(token))
}

function policyStatus(decision: PolicyDecision) {
  if (decision.code === 'PERMISSION_DENIED' || decision.code === 'SCOPE_DENIED') return 403
  if (decision.code === 'MODULE_UNLICENSED') return 402
  if (decision.code === 'RECORD_STATUS_BLOCKED' || decision.code === 'RECORD_DELETED') return 409
  return 403
}
